README.md
# nafsdm
**:warning: Current status: DEPRECATED, version 1.3.2-stable**

:warning: **nafsdm is no longer maintained and will not recieve any future updates! Use at your own risk.**

Manages DNS nodes and makes sure domains are saved in the slaves configs. Runs on Python and SSH.

*nafsdm stands for "not advanced, fast, simple dns manager"*

# Prerequisites & Compatibility
Before installing nafsdm, make sure you have at least one bind master and one slave already configured. The master needs to have SSH open (at least a firewall that only allows your slave IP's) and all slaves need to have bind already configured (**nafsdm DOES NOT install bind for you**)

nafsdm is tested to work on the following operating systems:

* Debian 8, 9 (7 should work fine)
* Ubuntu 16.04.2 (old versions like 14.04 should also work fine, but may lack systemd)
* CentOS 7

nafsdm currently works with:

* bind9 (or 'named' on CentOS)

# Installation
Installing nafsdm can be done using the guide below.

Before installing, make sure you do not have any existing installation of nafsdm (the installer will not upgrade current installations)

## Master installation
Connect to your master server and initiate the installation script.

`bash <(curl -s https://raw.githubusercontent.com/MrKaKisen/nafsdm/master/setupMaster.sh)`

Follow the steps in the installation.

If the above command fails, you can also manually download the script and execute it (using for example wget).

Once it's finished, run the master once, as it will generate keys (the installer also reminds you of this).

`nafsdm-master`

Now, copy the SSH key contents somewhere to your computer as it will be needed on the slaves later on.

`cat /home/master-nafsdm/.ssh/nafsdm_rsa` (this will print the key)

The master installation is now finished and should be ready to use. You can use `nafsdmctl` to add/remove domains or check the status of connected daemons.

## Slave installation
You will have to perform these steps on every slave you would like to install this on.
Connect to your slave and initiate the installation script.

`bash <(curl -s https://raw.githubusercontent.com/MrKaKisen/nafsdm/master/setupSlave.sh)`

Follow the steps in the installation.

Once it's finished, you can open up the config with your editor of choice (example below uses nano) to set it up correctly.

`nano /home/slave-nafsdm/config.conf`


Here is an explanation of what every option is.

`host` = hostname / IP for the master node we confgiured earlier

`user` = the username where nafsdm on the master installed. By default, it's master-nafsdm.

`update_interval` = how often the slave will update it's configuration (seconds).

`type` = type of the slave system. Is it ubuntu, debian or centos?

`bindPath` = path of configuration file nafsdm will write to (bind configuration)

`nodeName` = name of this node

The development section isn't something you should edit unless you're absolutely sure of what you're doing.

Once you've updated the config, you can paste over the SSH key you saved earlier to the slave. Paste it in to the file mentioned below.

`nano /home/slave-nafsdm/.ssh/master_key`

The key also needs to have correct permissions.

`chmod 600 /home/slave-nafsdm/.ssh/master_key` (nafsdm will **NOT** work if the correct permissions are not used)

You're done! You should now be able to start the slave (if everything is correctly configured).

`nafscli start`

As of version 1.2.5, you can control most of the basic operations on the slave using nafscli. Run nafscli for a list of available commands.

Replace start with stop or restart if you would like to do that later on (or status to check if it's running)

If anything fails, you can check the log.

`nafscli log`

When running the slave the first time, you will probably have to accept the "fingerprint". To accept it, stop the daemon.

`nafscli stop`

And run the slave in debug mode, which will let you accept the fingerprint.

`python /home/slave-nafsdm/pythondaemon/__main__.py`

When it asks you, just type "yes" and hit enter. When it's done, hit Control+C to stop the daemon and then boot the daemon back up again using service (as usual). It should look something like this:

`The authenticity of host 'example.example (0.0.0.0)' can't be established.`

`ECDSA key fingerprint is SHA256:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.`

`Are you sure you want to continue connecting (yes/no)?`

# Upgrading

To upgrade, please refer to the following wiki page: <https://github.com/MrKaKisen/nafsdm/wiki/Upgrading>

# Contributing
I gladly accept any pull requests that looks good! Just make sure you're working in the "development" branch and that you state in the request what you've modified and why. And don't create any new bugs :).

# Author
Email: contact@mrkakisen.net

nafsdm - Fast & easy DNS node manager for bind
Copyright (C) 2018 Vilhelm Prytz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
Version 1.3.2 (Status: RELEASED)
* Deprecation notice

Version 1.3.1 (Status: RELEASED)
* Bugfix: Slaves reported incorrect date to nafsdmctl slavestatus (minute instead of month under the "latest connection date" section)
* Slaves now check if another nafsdm-slave process is running before boot
* Fix issue #22
* Visual adjustments to the upgrade scripts.
* (minor): Slaves properly catch SIGINT and SIGTERM to exit gracefully
* (minor): Updated nafsdm-slave systemd service to only start after internet connection is available
* (minor): CHANGELOG is now included in install

Version 1.3 (Status: RELEASED)
* Upgrade scripts now check for "github_branch" to download correct latest version file.
* Upgrade scripts now return different exit codes depending on issue (dev versions will no longer crash the daemon due to unsupported version)
* Complete visual overhaul of nafsdmctl (looks a lot nicer)
* New development mode, allowing latest commit to always be installed on master and slave (not set version numbers).
* Slave: development functions are now enabled using the config.conf instead of adding files.
* Developer mode introduced in version 1.2.5-stable has been renamed to "skipVersionCheck"
* nafsdmctl: now has remove by domain and remove by ID functions
* nafsdmctl: now has status/start/stop/restart functions for the webinterface
* Slaves now report last connection time to master (nafsdmctl can show a table of which slaves has connected recently)
* nafscli logviewer no longer throws a traceback message on exit (using CTRL+C)
* Rewrote most log messages for both master and slave
* Ability to enable or disable nafsdm upgrades on start (if turned off, you will have to use nafscli upgrade) on slave daemon
* nafscli status now has -a argument which shows the full systemd status of the nafsdm daemon
* (minor): Update all copyright notices to 2018

Version 1.2.5 (Status: RELEASED)
* Added nafscli for the slave, a command-line interface
* Added developer mode for both Slave and Master (disables version checking)

Version 1.2.4 (Status: RELEASED)
* experimental: webinterface for the master, allowing control from a browser
* Slaves now check if their version matches with the one the master is running
Smaller changes/bugfixes:
* Remove double import of sys in nafsdmctl
* Added extra script to check if nafsdm is running (should work with monitoring software like nagios)
* Nicer looking setup scripts
* Installer now installs pip libs using a requirements.txt file (and updater).

Version 1.2.3 (Status: RELEASED)
* nafsdm-master now uses SQL to save domains (updated nafsdmctl and slave with the new save format)
* nafsdmctl list now prints in a nicer format
(from dev2)
* Bugfixes for dev1 issues (which need to be retested, therefore new dev release)

Version 1.2.2 (Status: RELEASED)
* Reworked logger function completely (now catches exceptions, new logging levels, stdout and filehandler)
* Small changes

Version 1.2.1 (Status: RELEASED)
* Development feature: specifiy GitHub branch to use
* Issue #8 fixed (nafsdm-slave not start on boot)
* Issue #11 fixed (nafsdmctl editor open causing slaves to crash).
* Issue #12 fixed (CTRL+C in nafsdmctl edit).
(from dev2)
* Remove non-working fix for issue #8.
* New fix for issue #8.
(from dev3, not released we just go straight to stable)
* New fix for issue #8
* Issue #11 silly fix

Version 1.2 (Status: RELEASED)
* Added DNSSEC support.
* Added upgrade scripts.
* Added support for automatic upgrades (1.2 and forwards)
* New edit function for nafsdmctl (master).
(from 1.2-dev3):
* Tons of bugfixes from 1.2-dev2. See release notes.

Version 1.1 (Status: RELEASED)
* New version files.
* New config parsing.
* Setup scripts has new version selection (and another fix on that since 1.1-dev2)
* Minor bug/spell fixes in master aswell as slave config changed (section name)

Version 1.0.1-stable (Status: RELEASED)
* Slave: Fixed serious writing issue where ' is used instead of ".
* Slave: Updated hostname finding function.

Version 1.0 (Status: RELEASED, now UNSTABLE)
* First version.


scripts/upgradeMaster.py
# nafsdm
# upgrade script for nafsdm master
# (c) Vilhelm Prytz 2018
# https://github.com/mrkakisen/nafsdm

# why have a seperate script for this? incase the thing needs more args than it knows

# imports
import subprocess
import os.path
import sys

# by default, path is manager
managerPath = "manager"

# import version (has changed in 1.4, we need to check which folder the uses has their version in)
if os.path.exists("/home/master-nafsdm/pythondaemon"):
    managerPath = "pythondaemon"
    sys.path.insert(0, "/home/master-nafsdm/pythondaemon")
    from version import version
else:
    sys.path.insert(0, "/home/master-nafsdm/manager")
    from version import version

def initUpgrade(github_branch="master", dev_ic_mode=False):
    try:
        output = subprocess.check_output(["/bin/bash", "/home/master-nafsdm/" + managerPath + "/tempUpgrade/temp_upgrade.sh", github_branch, str(dev_ic_mode)])
    except subprocess.CalledProcessError as grepexc:
        # we want grepexc to determine the exit code (if it's an issue or just unsupported version)
        exit_code = grepexc.returncode

        # these versions cannot tell the difference between unsupported and exception and has therefore their own section in this script
        if version == "1.2.5-stable" or version == "1.2.4-stable" or version == "1.2.3-stable" or version == "1.2.2-stable" or version == "1.2.1-stable" or version == "1.2-stable" or version == "1.1-stable" or version == "1.0.1-stable":
            return "exception"
        else:
            # return different depending on exit code (for versions after 1.2.5-stable)
            if int(exit_code) == 128:
                return "unsupported"
            if int(exit_code) == 1:
                return "exception"
            else:
                return "unknownException"

    return output
